sky/backends/backend_utils.py

```diff
@@ -1131,14 +1136,39 @@ def query_head_ip_with_retries(cluster_yaml: str, max_attempts: int = 1) -> str:
     return head_ip
 
 
+def get_stable_cluster_internal_ips(
+        handle: backends.Backend.ResourceHandle) -> List[str]:
+    if handle.stable_cluster_internal_ips is not None:
+        return handle.stable_cluster_internal_ips
+
+    cluster_external_ips = get_node_ips(handle.cluster_yaml,
+                                        handle.launched_nodes,
+                                        handle=handle)
+    cluster_internal_ips = get_node_ips(handle.cluster_yaml,
+                                        handle.launched_nodes,
+                                        handle=handle,
+                                        get_internal_ips=True)
+    internal_external_ips = list(zip(cluster_internal_ips,
+                                     cluster_external_ips))
+
+    # Ensure head node is the first element, then sort based on the external
+    # IPs for stableness
+    stable_internal_external_ips = [internal_external_ips[0]] + sorted(
+        internal_external_ips[1:], key=lambda x: x[1])
+    handle.stable_cluster_internal_ips = [
```

Comment:
Do we want to cache both internal and external IPs, so that in the `cloud_vm_ray_backend`, we can use those ips to run `setup` or `execute`?